### PR TITLE
Data-configuration: introduced ingress/egress data slices

### PR DESCRIPTION
**Pull request by jordigilh**
Introduces an ingress/egress slice for data paths in the workload's data configuration.
* Egress data path: sync from device to remote object storage.
* Ingress data path: sync from remote object storage to device. No disk capacity checks in the device implemented at the moment, but worth considering.
* Depends on https://github.com/project-flotta/flotta-operator/pull/235 first before this one is merged.

Even though the PR is on hold due to dependencies to the swagger changes in the flotta operator, the code changes that handle the new flows are implemented, so feel free to review at your discretion it while the other PR is being considered.

Note that the sync process does not delete data at the target destination when the source gets deleted. This is already by configuration for S3 syncs, not something included in this PR.

### CODE REVIEW

File: internal/datatransfer/monitor.go [hunk 1]
```diff
@@ -257,3 +231,44 @@ func (m *Monitor) Update(configuration models.DeviceConfigurationMessage) error
...
 	return nil
 }
+
+func containsDataPaths(dc *models.DataConfiguration) bool {
+	return dc != nil &&
+		(dc.Egress != nil && len(dc.Egress) > 0 ||
+			dc.Ingress != nil && len(dc.Ingress) > 0)
+}
+
+func syncDataConfiguration(syncWrapper FileSync, dataConfig *models.DataConfiguration, hostPath string) error {
```
Comment (jakub-dzon) [inline internal/datatransfer/monitor.go:241]: It syncs data, not the the configuration; please rename.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: internal/datatransfer/monitor.go [hunk 2]
```diff
@@ -196,41 +182,29 @@ func (m *Monitor) syncPaths() error {
...
-		}
-
-		if success {
-			m.storeLastUpdateTime(wd.Name)
+		err := syncDataConfiguration(syncWrapper, dataConfig, hostPath)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+			continue
 		}
+		m.storeLastUpdateTime(wd.Name)
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply internal/datatransfer/monitor.go:207]: I'm not following you here. If I'm not mistaken this line won't execute if there is an error in the data sync from line 202. Isn't it the same behavior as in the existing code?.

[the remark this answers is not in the document]
Comment (jordigilh) [reply internal/datatransfer/monitor.go:207]: Ok I get you know :)

File: internal/datatransfer/monitor.go [hunk 3]
```diff
@@ -107,33 +107,19 @@ func (m *Monitor) syncPathsWorkload(workloadName string) {
...
-		log.Debugf("synchronizing [device]%s => [remote]%s", source, target)
-
-		if err := syncWrapper.SyncPath(source, target); err != nil {
-			log.Errorf("error while synchronizing [device]%s => [remote]%s: %v", source, target, err)
-			success = false
-		}
-	}
-	if success {
+	err = syncDataConfiguration(syncWrapper, dataConfig, hostPath)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply internal/datatransfer/monitor.go]: done

File: internal/datatransfer/monitor.go (hunk 1 above)
Comment (jordigilh) [reply internal/datatransfer/monitor.go]: renamed to `syncData`

File: vendor/github.com/project-flotta/flotta-operator/models/data_configuration.go [hunk 4]
```diff
@@ -18,15 +18,22 @@ import (
 // swagger:model data-configuration
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply vendor/github.com/project-flotta/flotta-operator/models/data_configuration.go:18]: updated `go.mod` to reference the new code. This file is now included as part the `go.mod` commit change.

**Review by jakub-dzon — APPROVED** [review]

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Review by masayag — APPROVED** [review]

**Pull request merged**